pkg/storage/rulestore/config.go

```diff
@@ -0,0 +1,122 @@
+package rulestore
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/configs/client"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/azure"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/filesystem"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/gcs"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/s3"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/swift"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	ConfigDB = "configdb"
+
+	name   = "rule-store"
+	prefix = "rule-store."
+)
+
+var (
+	supportedBackends = []string{bucket.S3, bucket.GCS, bucket.Azure, bucket.Swift, bucket.Filesystem, ConfigDB}
+
+	ErrUnsupportedStorageBackend = errors.New("unsupported storage backend")
+)
+
+// Config configures a rule store.
+type Config struct {
+	bucket.Config `yaml:",inline"`
+	ConfigDB      client.Config `yaml:"configdb"`
+}
+
+// RegisterFlags registers the backend storage config.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.ConfigDB.RegisterFlagsWithPrefix(prefix, f)
+
+	cfg.S3.RegisterFlagsWithPrefix(prefix, f)
+	cfg.GCS.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Azure.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Swift.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Filesystem.RegisterFlagsWithPrefix(prefix, f)
+
+	f.StringVar(&cfg.Backend, prefix+"backend", "s3", fmt.Sprintf("Backend storage to use. Supported backends are: %s.", strings.Join(supportedBackends, ", ")))
+}
+
+func (cfg *Config) Validate() error {
+	if !util.StringsContain(supportedBackends, cfg.Backend) {
+		return ErrUnsupportedStorageBackend
+	}
+
+	if cfg.Backend == bucket.S3 {
```

Comment:
We shouldn't duplicate the bucket config `Validate()` but use it. I understanding you did it because the `Backend` validation wouldn't pass if it's configured `configdb`.

An option:
1. Create `bucket.Config.ValidateWithExtraBackends(extras []string)`
2. Pass `[]string{configdb}` to `ValidateWithExtraBackends()`
3. This function just calls `cfg.Bucket.Validate()`

Reply:
Makes sense. The only thing I would say is wouldn't we need `cfg.Bucket.Validate()` to call `cfg.Bucket. ValidateWithExtraBackends([]string{})` instead of vice versa?